File: backend_python/app/services/notification_service.py
"""
Notification service - mirrors INotificationService
"""
from app.repositories.firestore_repository import FirestoreRepository
from app.config import settings
from firebase_admin import messaging
from typing import Optional
import logging

logger = logging.getLogger(__name__)


class NotificationService:
    """
    Service for FCM notification operations.
    Mirrors the .NET NotificationService.
    Uses Firebase Admin SDK for FCM (more aligned with .NET FCM.Net).
    """
    
    def __init__(self, firestore_repo: FirestoreRepository):
        self.firestore_repo = firestore_repo
    
    async def send_notification(
        self,
        token: str,
        title: str,
        body: str,
        data: Optional[dict[str, str]] = None
    ) -> None:
        """
        Send FCM notification to a token.
        Mirrors SendNotificationAsync.
        """
        try:
            message = messaging.Message(
                notification=messaging.Notification(
                    title=title,
                    body=body
                ),
                data=data or {},
                token=token
            )
            
            response = messaging.send(message)
            logger.info(f"Notification sent successfully: {response}")
        except Exception as e:
            # Log error but don't throw (matches .NET behavior)
            logger.error(f"Error sending notification: {e}")
    
    async def send_notification_to_team(
        self,
        team_id: str,
        title: str,
        body: str,
        data: Optional[dict[str, str]] = None
    ) -> None:
        """
        Send FCM notification to team captain.
        Mirrors SendNotificationToTeamAsync.
        """
        try:
            # Get team captain's FCM token from users collection
            team = await self.firestore_repo.get_document("teams", team_id)
            if not team or "captainUid" not in team:
                logger.warning(f"Team {team_id} not found or missing captainUid")
                return
            
            captain_uid = team["captainUid"]
            if not captain_uid:
                logger.warning(f"Team {team_id} has empty captainUid")
                return
            
            user = await self.firestore_repo.get_document("users", captain_uid)
            if not user or "fcmToken" not in user:
                logger.warning(f"User {captain_uid} not found or missing fcmToken")
                return
            
            token = user["fcmToken"]
            if token:
                await self.send_notification(token, title, body, data)
        except Exception as e:
            logger.error(f"Error sending notification to team {team_id}: {e}")

File: backend/CricketTournament.Application/Services/NotificationService.cs
using CricketTournament.Application.Interfaces;
using CricketTournament.Domain.Interfaces;
using FCM.Net;
using Microsoft.Extensions.Configuration;

namespace CricketTournament.Application.Services;

public class NotificationService : INotificationService
{
    private readonly IFirestoreRepository _firestoreRepository;
    private readonly IConfiguration _configuration;
    private readonly Server _fcmServer;

    public NotificationService(IFirestoreRepository firestoreRepository, IConfiguration configuration)
    {
        _firestoreRepository = firestoreRepository;
        _configuration = configuration;
        var serverKey = _configuration["FCM:ServerKey"];
        _fcmServer = new Server(serverKey ?? string.Empty);
    }

    public async Task SendNotificationAsync(string token, string title, string body, Dictionary<string, string>? data = null)
    {
        try
        {
            var message = new Message
            {
                To = token,
                Notification = new FCM.Net.Notification
                {
                    Title = title,
                    Body = body
                },
                Data = data ?? new Dictionary<string, string>()
            };

            await _fcmServer.SendAsync(message);
        }
        catch (Exception ex)
        {
            // Log error but don't throw
            Console.WriteLine($"Error sending notification: {ex.Message}");
        }
    }

    public async Task SendNotificationToTeamAsync(string teamId, string title, string body, Dictionary<string, string>? data = null)
    {
        // Get team captain's FCM token from users collection
        var team = await _firestoreRepository.GetDocumentAsync<Dictionary<string, object>>("teams", teamId);
        if (team == null || !team.ContainsKey("captainUid")) return;

        var captainUid = team["captainUid"].ToString();
        if (string.IsNullOrEmpty(captainUid)) return;

        var user = await _firestoreRepository.GetDocumentAsync<Dictionary<string, object>>("users", captainUid);
        if (user == null || !user.ContainsKey("fcmToken")) return;

        var token = user["fcmToken"].ToString();
        if (!string.IsNullOrEmpty(token))
        {
            await SendNotificationAsync(token, title, body, data);
        }
    }
}

